### PR TITLE
Core: allow for a single player all_state

### PR DESCRIPTION
**Pull request by alwaysintreble**
## What is this fixing or adding?
allows a world to provide a player to `multiworld.get_all_state()`, building an all_state that only has the items and sweeps the locations of that player. This also fixes an existing bug with `get_all_state` where it returns the actual state the first time it gets built instead of a copy of it, allowing extra items to be added to the pool.

## How was this tested?
hours of debugging 😔 
Tested in combination with #2415 and this gave an additional ~25% speed increase. for 300 worlds using pre_fill methods

### CODE REVIEW

**Comment by NewSoupVi** [comment]
Question: When you say "this gave an additional ~25% increase", that means you went and changed some `get_all_state`s to `get_all_state(player)` - But that isn't included in this PR because you moreso wanted to get a feel for it and not actually touch other people's worlds?

If so, this is probably worth a world maintainer ping in the Discord or something, like "please switch to get_all_state(player) if your implementation allows it"

File: BaseClasses.py [hunk 1]
```diff
@@ -367,23 +367,37 @@ def get_entrance(self, entrance_name: str, player: int) -> Entrance:
...
+            for player in self.player_ids:
+                subworld = self.worlds[player]
+                for item in subworld.get_pre_fill_items():
+                    subworld.collect(ret, item)
+        ret.sweep_for_events(locations=self.get_filled_locations(player))  # TODO player=player
 
         if use_cache:
-            self._all_state = ret
+            if player:
+                setattr(self, f"_all_state{player}", None)
```
Comment (NewSoupVi) [inline BaseClasses.py:397]: Wait what?

Comment (alwaysintreble) [reply BaseClasses.py:397]: lmao oops. i was testing it with a cache then temp removed it to find a bug and put it back wrong

**Comment by alwaysintreble** [comment]
> Question: When you say "this gave an additional ~25% increase", that means you went and changed some `get_all_state`s to `get_all_state(player)` in some worlds - But that isn't included in this PR because you moreso wanted to get a feel for it and not actually touch other people's worlds?

i tested it under the same conditions as the linked PR

**Pull request closed without merging**